packages/datetime/src/dateRangePicker.tsx

```diff
@@ -324,29 +346,86 @@ export class DateRangePicker
-        } else if (start == null && end != null) {
-            nextValue = this.createRange(day, end);
-        } else {
-            const isStart = DateUtils.areSameDay(start, day);
-            const isEnd = DateUtils.areSameDay(end, day);
-            if (isStart && isEnd) {
-                nextValue = [null, null];
-            } else if (isStart) {
-                nextValue = [null, end];
-            } else if (isEnd) {
-                nextValue = [start, null];
+        const { allowSingleDayRange } = this.props;
+
+        // rename for conciseness
+        const boundary = this.props.preferredBoundaryToModify;
+
+        if (boundary != null) {
+            const boundaryDate = (boundary === DateRangeBoundary.START) ? start : end;
+            const otherBoundaryDate = (boundary === DateRangeBoundary.START) ? end : start;
+
+            if (boundaryDate == null && otherBoundaryDate == null) {
+                nextValue = this.createRangeForBoundary(day, null, boundary);
+            } else if (boundaryDate != null && otherBoundaryDate == null) {
+                const nextBoundaryDate = DateUtils.areSameDay(boundaryDate, day) ? null : day;
+                nextValue = this.createRangeForBoundary(nextBoundaryDate, null, boundary);
+            } else if (boundaryDate == null && otherBoundaryDate != null) {
+                if (DateUtils.areSameDay(day, otherBoundaryDate)) {
+                    const nextOtherBoundaryDate = allowSingleDayRange ? otherBoundaryDate : null;
+                    nextValue = this.createRangeForBoundary(day, nextOtherBoundaryDate, boundary);
+                } else if (this.isDateOverlappingOtherBoundary(day, otherBoundaryDate, boundary)) {
+                    nextValue = this.createRangeForBoundary(otherBoundaryDate, day, boundary);
+                } else {
+                    nextValue = this.createRangeForBoundary(day, otherBoundaryDate, boundary);
+                }
             } else {
+                // both boundaryDate and otherBoundaryDate are already defined
+                if (DateUtils.areSameDay(boundaryDate, day)) {
+                    const isSingleDayRangeSelected = DateUtils.areSameDay(boundaryDate, otherBoundaryDate);
+                    const nextOtherBoundaryDate = isSingleDayRangeSelected ? null : otherBoundaryDate;
+                    nextValue = this.createRangeForBoundary(null, nextOtherBoundaryDate, boundary);
+                } else if (DateUtils.areSameDay(day, otherBoundaryDate)) {
+                    const nextOtherBoundaryDate = (allowSingleDayRange) ? otherBoundaryDate : null;
+                    nextValue = this.createRangeForBoundary(day, nextOtherBoundaryDate, boundary);
+                } else if (this.isDateOverlappingOtherBoundary(day, otherBoundaryDate, boundary)) {
+                    nextValue = this.createRangeForBoundary(day, null, boundary);
+                } else {
+                    // extend the date range with an earlier boundaryDate date
```

Comment:
nit: `earlier` is a relic from older code. reword this to indicate we're "stretching" the range (or something).
nit: `boundaryDate date` => `boundaryDate`